### PR TITLE
feat: reverse_proxy binds 8000 on-demand with devserver, lazy image pulls, and stop instead of destroy on down

### PR DESCRIPTION
**Pull request by untitaker**
Some ppl (most of them being reviewers for this PR) had issues with sentry now permanently listening on port 8000. This should help alleviate this concern.

Into your `sentry.conf.py`:

```
# Start nginx container as part of `sentry devserver`.
SENTRY_DEVSERVICES['reverse_proxy']['with_devserver'] = True
```

I also use this to lazily setup and teardown symbolicator and snuba. You'd think this is slow but it isn't.

I'd merge this, let the reviewers run it for a while and if it works for them make it the default for at least `reverse_proxy`.

### CODE REVIEW

**Comment by dcramer** [comment]
didnt review the implementation but i like the idea

aside: could we make this an external library that we might reuse it in our other projects?

File: src/sentry/runner/commands/devservices.py [hunk 1]
```diff
@@ -48,11 +50,64 @@ def devservices():
...
+
+    from sentry.runner import configure
+
+    configure()
+
+    client = get_docker_client()
+    containers = _prepare_containers(project)
+    container = _start_service(client, service, containers, project, fast=True)
+
+    def exit_handler(*_):
```
Comment (joshuarli) [inline src/sentry/runner/commands/devservices.py:79]: Why `*_`? Fairly sure that's not needed.

Comment (untitaker) [reply src/sentry/runner/commands/devservices.py:79]: signal handlers take `signum, frame` while atexit takes nothing

File: src/sentry/runner/commands/devservices.py [hunk 2]
```diff
@@ -48,11 +50,64 @@ def devservices():
...
+        click.echo(line, nl=False)
+
+
 @devservices.command()
 @click.option("--project", default="sentry")
 @click.option("--exclude", multiple=True, help="Services to ignore and not run.")
-def up(project, exclude):
-    "Run/update dependent services."
+@click.argument("service", nargs=1, required=False)
+def up(project, service, exclude):
```
Comment (joshuarli) [inline src/sentry/runner/commands/devservices.py:98]: `service` isn't actually used in the function body.

```
joshua.li@tim-apple sentry feat/devservices-with-devserver $ sentry devservices up redis
19:30:50 [WARNING] sentry.utils.geo: settings.GEOIP_PATH_MMDB not configured.
19:30:53 [INFO] sentry.plugins.github: apps-not-configured
! Skipping bigtable due to only_if condition
! Skipping memcached due to only_if condition
> Creating 'sentry_postgres' container (listening: ('127.0.0.1', 5432))
> Pulling image 'us.gcr.io/sentryio/relay:latest'
...
```

Comment (untitaker) [reply src/sentry/runner/commands/devservices.py]: you're right this was dead code from an earlier experiment

File: src/sentry/runner/commands/devservices.py (hunk 1 above)
Comment (joshuarli) [reply src/sentry/runner/commands/devservices.py:79]: Oh sorry, overlooked the fact you were reusing the same handler for both atexit and signals. For context, I'm used to doing something like:

```
def sig_exit_handler(signo, frame):
    exit_handler()
    sys.exit(128 + signo)  # http://tldp.org/LDP/abs/html/exitcodes.html
```

File: src/sentry/runner/commands/devservices.py (hunk 2 above)
Comment (joshuarli) [reply src/sentry/runner/commands/devservices.py]: Surprised flake8 doesn't catch this.

**Comment by joshuarli** [comment]
At this point it works like this:

- Configure `SENTRY_DEVSERVICES['reverse_proxy']['with_devserver'] = True`
- `sentry devservices up`
  - reverse_proxy will occupy 8000
- `sentry devserver`
  - works, code to start reverse_proxy noops
- send SIGINT to devserver
  - reverse_proxy will be taken down along with devserver, freeing 8000

I'm going to attempt to polish this by not occupying 8000 in the first place; it should only be done on demand with devserver.

**Comment by joshuarli** [comment]
Just thinking aloud here: it would maybe be nice to add a `--fast` option to devservices up and down to prevent the destroy + pull happening every time for devservices marked with pull True. This would be trading speed for correctness.

**Comment by joshuarli** [comment]
@untitaker Thoughts on this patch? https://gist.github.com/joshuarli/51f5155488a757eb49e494431179cdc0

Makes stuff faster by just stopping the containers on down, and devservices with pull True will be stopped and recreated on up:

```
joshua.li@tim-apple commands * feat/devservices-with-devserver $ sentry devservices down
> Stopping 'sentry_clickhouse' container
> Stopping 'sentry_redis' container
> Stopping 'sentry_symbolicator' container
> Stopping 'sentry_snuba' container
> Stopping 'sentry_kafka' container
> Stopping 'sentry_zookeeper' container
> Stopping 'sentry_reverse_proxy' container
> Stopping 'sentry_relay' container
> Stopping 'sentry_postgres' container
joshua.li@tim-apple commands * feat/devservices-with-devserver $ docker ps
CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS              PORTS               NAMES
joshua.li@tim-apple commands * feat/devservices-with-devserver $ docker
joshua.li@tim-apple commands * feat/devservices-with-devserver $ sentry devservices up
22:34:38 [WARNING] sentry.utils.geo: settings.GEOIP_PATH_MMDB not configured.
22:34:42 [INFO] sentry.plugins.github: apps-not-configured
! Skipping bigtable due to only_if condition
! Skipping memcached due to only_if condition
> Starting existing 'sentry_postgres' container (listening: ('127.0.0.1', 5432))
> Creating 'sentry_relay' container (listening: ('127.0.0.1', 3000))
> Starting existing 'sentry_reverse_proxy' container (listening: ('127.0.0.1', 8000))
> Starting existing 'sentry_zookeeper' container
> Starting existing 'sentry_kafka' container (listening: ('127.0.0.1', 9092))
> Creating 'sentry_snuba' container (listening: ('127.0.0.1', 1218))
> Creating 'sentry_symbolicator' container (listening: ('127.0.0.1', 3021))
> Starting existing 'sentry_redis' container (listening: ('127.0.0.1', 6379))
> Starting existing 'sentry_clickhouse' container (listening: ('127.0.0.1', 8123), ('127.0.0.1', 9009), ('127.0.0.1', 9000))
```

**Comment by joshuarli** [comment]
@untitaker I think I've done it, see branch https://github.com/getsentry/sentry/tree/feat/devservices-with-devserver-joshuarli.

This is what the experience looks like.

Fresh slate (after completely nuking all my docker containers + images)

```
$ sentry devservices up
! Skipping bigtable due to only_if condition
! Skipping memcached due to only_if condition
> Pulling image 'postgres:9.6-alpine'
> Creating 'sentry_postgres' container (listening: ('127.0.0.1', 5432))
> Pulling image 'us.gcr.io/sentryio/relay:latest'
> Creating 'sentry_relay' container (listening: ('127.0.0.1', 3000))
> Pulling image 'nginx:1.16.1'
> Creating 'sentry_reverse_proxy' container (listening: ('127.0.0.1', 8000))
> Not starting container 'sentry_reverse_proxy' because it should be started on-demand with devserver.
> Pulling image 'confluentinc/cp-zookeeper:5.1.2'
> Creating 'sentry_zookeeper' container
> Pulling image 'confluentinc/cp-kafka:5.1.2'
> Creating 'sentry_kafka' container (listening: ('127.0.0.1', 9092))
> Pulling image 'getsentry/snuba:latest'
> Creating 'sentry_snuba' container (listening: ('127.0.0.1', 1218))
> Pulling image 'us.gcr.io/sentryio/symbolicator:latest'
> Creating 'sentry_symbolicator' container (listening: ('127.0.0.1', 3021))
> Pulling image 'redis:5.0-alpine'
> Creating 'sentry_redis' container (listening: ('127.0.0.1', 6379))
> Pulling image 'yandex/clickhouse-server:19.11'
> Creating 'sentry_clickhouse' container (listening: ('127.0.0.1', 8123), ('127.0.0.1', 9009), ('127.0.0.1', 9000))

$ lsof -nP -i4TCP:8000  # no output; unoccupied
```

**Comment by joshuarli** [comment]
@untitaker I think I've done it, see branch https://github.com/getsentry/sentry/tree/feat/devservices-with-devserver-joshuarli.

This is what the experience looks like.

Fresh slate (after completely nuking all my docker containers + images):

```
$ sentry devservices up
! Skipping bigtable due to only_if condition
! Skipping memcached due to only_if condition
> Pulling image 'postgres:9.6-alpine'
> Creating 'sentry_postgres' container (listening: ('127.0.0.1', 5432))
> Pulling image 'us.gcr.io/sentryio/relay:latest'
> Creating 'sentry_relay' container (listening: ('127.0.0.1', 3000))
> Pulling image 'nginx:1.16.1'
> Creating 'sentry_reverse_proxy' container (listening: ('127.0.0.1', 8000))
> Not starting container 'sentry_reverse_proxy' because it should be started on-demand with devserver.
> Pulling image 'confluentinc/cp-zookeeper:5.1.2'
> Creating 'sentry_zookeeper' container
> Pulling image 'confluentinc/cp-kafka:5.1.2'
> Creating 'sentry_kafka' container (listening: ('127.0.0.1', 9092))
> Pulling image 'getsentry/snuba:latest'
> Creating 'sentry_snuba' container (listening: ('127.0.0.1', 1218))
> Pulling image 'us.gcr.io/sentryio/symbolicator:latest'
> Creating 'sentry_symbolicator' container (listening: ('127.0.0.1', 3021))
> Pulling image 'redis:5.0-alpine'
> Creating 'sentry_redis' container (listening: ('127.0.0.1', 6379))
> Pulling image 'yandex/clickhouse-server:19.11'
> Creating 'sentry_clickhouse' container (listening: ('127.0.0.1', 8123), ('127.0.0.1', 9009), ('127.0.0.1', 9000))

$ lsof -nP -i4TCP:8000  # no output; unoccupied
```

Fire up devserver:

```
$ sentry devserver
01:08:18 system        | webpack started (pid=56712)
01:08:18 system        | reverse_proxy started (pid=56713)
01:08:18 system        | ingest started (pid=56714)
01:08:18 system        | server started (pid=56715)
-- snip --
01:08:37 reverse_proxy | ! Skipping bigtable due to only_if condition
01:08:37 reverse_proxy | ! Skipping memcached due to only_if condition
01:08:37 reverse_proxy | > Creating 'sentry_reverse_proxy' container (listening: ('127.0.0.1', 8000))
01:08:37 server        | *** Starting uWSGI 2.0.18 (64bit) on [Wed Apr 22 01:08:37 2020] ***
-- snip --

$ lsof -nP -i4TCP:8000
COMMAND     PID      USER   FD   TYPE             DEVICE SIZE/OFF NODE NAME
com.docke 69761 joshua.li   58u  IPv4 0xe4a99cf72804f617      0t0  TCP 127.0.0.1:8000 (LISTEN)
```

Kill devserver:

```
^C 01:10:13 system        | SIGINT received
01:10:13 system        | sending SIGTERM to webpack (pid 56712)
01:10:13 system        | sending SIGTERM to reverse_proxy (pid 56713)
01:10:13 system        | sending SIGTERM to ingest (pid 56714)
01:10:13 system        | sending SIGTERM to server (pid 56715)
01:10:13 reverse_proxy | Shutting down reverse_proxy
01:10:13 server        | SIGINT/SIGQUIT received...killing workers...
01:10:13 system        | webpack stopped (rc=-15)
01:10:13 ingest        | 01:10:13 [INFO] batching-kafka-consumer: Partitions revoked: [TopicPartition{topic=ingest-attachments,partition=0,offset=-1001,error=None}, TopicPartition{topic=ingest-events,partition=0,offset=-1001,error=None}, TopicPartition{topic=ingest-transactions,partition=0,offset=-1001,error=None}]
01:10:13 reverse_proxy | Shutting down reverse_proxy
01:10:13 system        | reverse_proxy stopped (rc=0)
01:10:13 system        | ingest stopped (rc=0)
01:10:14 server        | worker 1 buried after 1 seconds
01:10:14 server        | goodbye to uWSGI.
01:10:14 system        | server stopped (rc=0)

$ lsof -nP -i4TCP:8000  # no output; unoccupied
```

Shut down devservices (This no longer deletes containers. Anything that needs to be re-pulled will be done smartly in up.)

```
$ sentry devservices down
> Stopping 'sentry_reverse_proxy' container
> Stopping 'sentry_clickhouse' container
> Stopping 'sentry_redis' container
> Stopping 'sentry_symbolicator' container
> Stopping 'sentry_snuba' container
> Stopping 'sentry_kafka' container
> Stopping 'sentry_zookeeper' container
> Stopping 'sentry_relay' container
> Stopping 'sentry_postgres' container
```

Start devservices:

Notice the "Starting existing" for all containers that belong to devservices without pull = True.

```
$ sentry devservices up
! Skipping bigtable due to only_if condition
! Skipping memcached due to only_if condition
> Starting existing 'sentry_postgres' container (listening: ('127.0.0.1', 5432))
> Pulling image 'us.gcr.io/sentryio/relay:latest'
> Creating 'sentry_relay' container (listening: ('127.0.0.1', 3000))
> Creating 'sentry_reverse_proxy' container (listening: ('127.0.0.1', 8000))
> Not starting container 'sentry_reverse_proxy' because it should be started on-demand with devserver.
> Starting existing 'sentry_zookeeper' container
> Starting existing 'sentry_kafka' container (listening: ('127.0.0.1', 9092))
> Pulling image 'getsentry/snuba:latest'
> Creating 'sentry_snuba' container (listening: ('127.0.0.1', 1218))
> Pulling image 'us.gcr.io/sentryio/symbolicator:latest'
> Creating 'sentry_symbolicator' container (listening: ('127.0.0.1', 3021))
> Starting existing 'sentry_redis' container (listening: ('127.0.0.1', 6379))
> Starting existing 'sentry_clickhouse' container (listening: ('127.0.0.1', 8123), ('127.0.0.1', 9009), ('127.0.0.1', 9000))

$ lsof -nP -i4TCP:8000  # no output; unoccupied
```

**Comment by untitaker** [comment]
Yes, sounds good!

**Comment by joshuarli** [comment]
@untitaker Okay, going to merge into this branch then.

Do you agree with making `SENTRY_DEVSERVICES['reverse_proxy']['with_devserver'] = True` default now? I don't see why not.

**Comment by untitaker** [comment]
yes but please don't change the default no other services for now. relay e.g. is broken if you flip the flag for it

**Comment by joshuarli** [comment]
> yes but please don't change the default no other services for now. relay e.g. is broken if you flip the flag for it

For posterity: this means `with_devserver` shouldn't be set on relay in the conf.

**Comment by jan-auer** [comment]
@untitaker can you elaborate why Relay would break? I was hoping to enable it specifically for relay to avoid the auth poll in the background when sentry is down.

**Review by joshuarli — APPROVED** [review]
Markus said it looks good. Gonna merge.

**Comment by untitaker** [comment]
@jan-auer relay would not break, but relay being down would cause nginx to not start because it tries to resolve `sentry_relay` hostname at startuo

**Comment by jan-auer** [comment]
@joshuarli This broke the `onprem` build since `import docker` is not available. Would it be possible to make this import optional or fix this in the build script?

**Comment by untitaker** [comment]
I can fix that, should be straightforward

**Pull request merged**